include/revng/Storage/StorageClient.h

```diff
@@ -0,0 +1,114 @@
+#pragma once
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include <memory>
+
+#include "llvm/ADT/StringRef.h"
+#include "llvm/Support/Error.h"
+#include "llvm/Support/FileSystem.h"
+#include "llvm/Support/MemoryBuffer.h"
+#include "llvm/Support/Path.h"
+
+#include "revng/Support/Assert.h"
+#include "revng/Support/Debug.h"
+
+namespace revng {
+
+class WritableFile {
+private:
+  std::unique_ptr<llvm::raw_pwrite_stream> OS;
+  std::function<llvm::Error()> Closer;
+  std::function<void()> Aborter;
+
+public:
+  WritableFile(std::unique_ptr<llvm::raw_pwrite_stream> &&OS,
+               std::function<llvm::Error()> &&Closer,
+               std::function<void()> &&Aborter) :
+    OS(std::move(OS)), Closer(std::move(Closer)), Aborter(std::move(Aborter)) {}
+  WritableFile(const WritableFile &Other) = delete;
+  WritableFile &operator=(const WritableFile &Other) = delete;
+  WritableFile(WritableFile &&Other) = default;
+  WritableFile &operator=(WritableFile &&Other) = default;
+  ~WritableFile() {
+    if (OS != nullptr)
+      Aborter();
+  }
+
+  llvm::raw_pwrite_stream &getOS() { return *OS; }
+
+  llvm::Error close() {
```

Comment:
`keep` is for when the file is finalized, I don't think anyone is used to write stuff after invoking `keep`.

If you prefer, we can call this guy and `StorageClient::finalizeWrites` `commit`, and in the doc we can say "remember to always call `commit` or it will be like you did nothing!" (even if that's not true for non-S3).

Reply:
Changed to `WritableFile::os` and `WritableFile::commit` and `ReadableFile::buffer` (for consistency)
Also renamed `StorageClient::finalizeWrites` to `StorageClient::commit`